lib/elixir/lib/exception.ex

```diff
@@ -1091,18 +1091,85 @@ defmodule Protocol.UndefinedError do
 end
 
 defmodule KeyError do
-  defexception [:key, :term]
+  defexception [:key, :term, :message]
 
   @impl true
-  def message(exception) do
-    msg = "key #{inspect(exception.key)} not found"
+  def message(exception = %{message: nil}), do: message(exception.key, exception.term)
+  def message(%{message: message}), do: message
+
+  def message(key, term) do
+    message = "key #{inspect(key)} not found"
 
-    if exception.term != nil do
-      msg <> " in: #{inspect(exception.term)}"
+    if term != nil do
+      message <> " in: #{inspect(term)}"
     else
-      msg
+      message
     end
   end
+
+  @impl true
+  def blame(exception = %{term: nil}, stacktrace) do
+    message = message(exception.key, exception.term)
+    {%{exception | message: message}, stacktrace}
+  end
+
+  def blame(exception = %{term: term}, stacktrace) when is_map(term) do
+    available_keys = Map.keys(term)
+    hint = did_you_mean(exception.key, available_keys)
+    message = message(exception.key, term) <> IO.iodata_to_binary(hint)
+    {%{exception | message: message}, stacktrace}
+  end
+
+  def blame(exception = %{term: term}, stacktrace) when is_list(term) do
+    available_keys = Keyword.keys(term)
```

Comment:
Unfortunately I think there is a bug here. `Keyword.keys` assumes a keyword is given. So if you do `Keyword.fetch!([1, foo: :bar], :bar)`, it will invoke this clause which then fails on Keyword.keys. My suggestion is to unify both clauses above and do something like this:

    if (is_map(term) or Keyword.keyword?(term)) and Enum.all?(term, fn {k, _} -> is_atom(k) end) do
      ...
    else
      {message, stacktrace}
    end

I know we can have mixed keys in maps, but I think we should start with a more conservative approach for now.

Reply:
Ah - I didn't think it was possible for a non-keyword list to be passed as the term to a `KeyError`. I'll update.